src/universal_transfer_operator/dataset.py

```diff
@@ -0,0 +1,306 @@
+import logging
+import warnings
+from typing import TYPE_CHECKING, Any, List
+from urllib.parse import unquote, urlparse
+
+from airflow.exceptions import AirflowException
+from airflow.providers_manager import ProvidersManager
+from airflow.utils.module_loading import import_string
+
+if TYPE_CHECKING:
+    from airflow.models.connection import Connection  # Avoid circular imports.
+
+from src.universal_transfer_operator.data_providers.base import DataProviders
+
+log = logging.getLogger(__name__)
+
+
+class Dataset:
+    r"""
+    The Dataset class refers to the pointer to the data and reference
+    to it along with the connection details.
+
+    The namespace for a dataset is the unique name for its dataset. The
+    namespace and name of a dataset can be combined to
+    form a URI (scheme:[//authority]path)
+    * Namespace = scheme:[//authority] (the Dataset)
+    * Name = path (the datasets)
+
+    Broadly, Datasets can be of following categories:
+    * Databases / Data Warehouses
+    * File System / Blob storage
+    * APIs
+
+    ### Databases / Data Warehouses:
+    Datasets are called tables. Tables are organised in databases and schemas.
+
+    #### Postgres:
+    Dataset hierarchy:
+     * Host
+     * Port
+
+    Naming hierarchy:
+     * Database
+     * Schema
+     * Table
+
+    Identifier:
+     * Namespace: postgres://{host}:{port} of the service instance.
+       * Scheme = postgres
+       * Authority = {host}:{port}
+     * Unique name: {database}.{schema}.{table}
+       * URI =  postgres://{host}:{port}/{database}.{schema}.{table}
+
+    #### MySQL:
+    Dataset hierarchy:
+     * Host
+     * Port
+
+    Naming hierarchy:
+     * Database
+     * Table
+
```

Comment:
Do we want to have this level of detail in the base Dataset class docstring?

Reply:
I was thinking if we use sphinx, this doc-string would be useful. I could also add these details in a separate doc-string.

From the user's perspective, I felt it's important to know how URI has to be passed for different providers. What do you think?